Rewrite `vello_cpu` to be more SIMD-friendly

LaurenzV opened the pull request:
This is a first attempt at rewriting `vello_cpu` to be more SIMD-friendly. 

As can be seen in the code, there are still quite a few TODOs left. The usage of SIMD types is also not very consistent (in some cases we are using f32x16, in some cases we are limited to f32x4), but it should be a good enough basis for any platform that provides 128-bit SIMD capabilities, which I think is a good start.

This PR should fix the performance regression that was introduced when adding the f32 pipeline. There are certain parts of the pipeline (namely opaque filling and alpha compositing) which only marginally improve as part of this PR on NEON because it already auto-vectorized very well before, but other parts like gradients, blurred rects and images (especially images with bilinear and bicubic filtering) are _much_ faster now.

It also incorporates the gradient LUT change from a previous PR.

I've adjusted a gradient test case so that it doesn't trigger the floating point impreciseness we have already had before.

LaurenzV commented:
@ajakubowicz-canva Any idea what the WASM issue is about?

LaurenzV commented:
(Assuming you have a device that supports NEON).

nicoburns commented:
@LaurenzV It seems that there was something fishy going on. And it was that I was still multiplying the alpha input to `push_layer` by `255` even though it's `f32` not `u8` now. For some reason that causes significant slowdown (on all of this branch, main, and 0.0.1) even though it successfully renders. I'm seeing very good results with this branch (but also decent with `main` now that I've fixed the alpha issue).

See: [#vello > vello_cpu performance @ 💬](https://xi.zulipchat.com/#narrow/channel/197075-vello/topic/vello_cpu.20performance/near/526825491) for more and some numbers from Blitz with this branch.

ajakubowicz-canva commented:
I think it's fine to include here to unblock CI.

LaurenzV commented:
This is with all of the SIMD operations necessary for image rendering already implemented?

ajakubowicz-canva commented on Cargo.toml:
```diff
@@ -115,7 +115,7 @@ rayon = { version = "1.10.0" }
 crossbeam-channel = "0.5.15"
 ordered-channel = { version = "1.2.0", features = ["crossbeam-channel"] }
-fearless_simd = { git = "https://github.com/raphlinus/fearless_simd", rev = "8c22f7c", default-features = false }
+fearless_simd = { git = "https://github.com/raphlinus/fearless_simd", rev = "8a1cea7", default-features = false }
```
https://github.com/raphlinus/fearless_simd/pull/25 can be merged and commit can be on main branch.

LaurenzV replied:
Sorry, I missed that yesterday!

LaurenzV commented:
No worries, wasn't expecting you to take a detailed look at everything! I understand it's hard to review, but splitting this up into multiple smaller PRs will be close to impossible I think. Will be addressing your comments and then merge.

LaurenzV commented on sparse_strips/vello_cpu/src/dispatch/single_threaded.rs:
```diff
@@ -124,12 +127,44 @@ impl Dispatcher for SingleThreadedDispatcher {
-                Self::rasterize::<f32>(self, buffer, width, height, encoded_paints);
-            }
+            RenderMode::OptimizeSpeed => match self.level {
+                #[cfg(all(feature = "std", target_arch = "aarch64"))]
+                Level::Neon(n) => {
+                    self.rasterize_with::<vello_common::fearless_simd::Neon, U8Kernel>(
+                        n,
+                        buffer,
+                        width,
+                        height,
+                        encoded_paints,
+                    );
+                }
+                _ => self.rasterize_with::<Fallback, U8Kernel>(
+                    Fallback::new(),
+                    buffer,
+                    width,
+                    height,
+                    encoded_paints,
+                ),
+            },
+            RenderMode::OptimizeQuality => match self.level {
+                #[cfg(all(feature = "std", target_arch = "aarch64"))]
+                Level::Neon(n) => {
+                    self.rasterize_with::<vello_common::fearless_simd::Neon, F32Kernel>(
+                        n,
+                        buffer,
+                        width,
+                        height,
+                        encoded_paints,
+                    );
+                }
+                _ => self.rasterize_with::<Fallback, F32Kernel>(
+                    Fallback::new(),
+                    buffer,
+                    width,
+                    height,
+                    encoded_paints,
+                ),
+            },
```
Sure, let me try to add it.

LaurenzV commented on sparse_strips/vello_cpu/src/fine/common/gradient/mod.rs:
```diff
@@ -0,0 +1,171 @@
+    if pad {
+        val.max(0.0).min(1.0)
+    } else {
+        (val - val.floor()).fract()
```
🤔 I think you are right. 😄

LaurenzV replied:
Ah, nevermind, no you can't because otherwise the value van be negative as well.

LaurenzV commented on sparse_strips/vello_sparse_tests/snapshots/gradient_linear_with_y_repeat.png:
Yeah.

LaurenzV commented on sparse_strips/vello_cpu/src/fine/common/image.rs:
```diff
@@ -0,0 +1,543 @@
+        // center of the location we are sampling, and sample those points
+        // using a cubic filter to weight each location's contribution.
+
+        #[inline(always)]
+        fn fract<S: Simd>(val: f32x4<S>) -> f32x4<S> {
+            val - val.floor()
+        }
```
This is not exactly the same, because it has different behavior for negative numbers. Will add a comment.

LaurenzV commented on sparse_strips/vello_cpu/src/fine/highp/mod.rs:
Might be worth exploring in the future!

LaurenzV commented on sparse_strips/vello_dev_macros/src/bench.rs:
```diff
@@ -35,15 +37,40 @@ pub(crate) fn vello_bench_inner(_: TokenStream, item: TokenStream) -> TokenStrea
+                });
+            }
+
+            // Uncomment this to enable f32_scalar benchmarks.
+            // c.bench_function(&get_bench_name(&#input_fn_name_str, "f32_scalar"), |b| {
+            //     run_float(b, vello_common::fearless_simd::Fallback::new());
+            // });
+
+            // Uncomment this to enable f32_neon benchmarks.
+            // #[cfg(target_arch = "aarch64")]
+            // if let Some(neon) = Level::new().as_neon() {
+            //     c.bench_function(&get_bench_name(&#input_fn_name_str, "f32_neon"), |b| {
+            //         run_float(b, neon);
+            //     });
+            // }
```
I mainly did it because performance-wise, we only really care about the u8 SIMD version. It was a bit annoying to always manually comment it out when running the benchmarks, so I just left it commented out. I could change it back, but I doubt that anyone would want to run benchmarks for the other versions.

LaurenzV commented on sparse_strips/vello_dev_macros/src/lib.rs:
```diff
@@ -15,7 +15,7 @@ mod test;
 // than 1. For example, if the target pixel is (233, 43, 64, 100), then permissible
 // values are (232, 43, 65, 101) or (233, 42, 64, 100), but not (231, 43, 64, 100).
-const DEFAULT_CPU_U8_TOLERANCE: u8 = 1;
+const DEFAULT_CPU_U8_TOLERANCE: u8 = 2;
```
If I change it back to 1 those are the failures I get:

<img width="522" alt="image" src="https://github.com/user-attachments/assets/2c40e420-655f-4043-83c0-e6067f1e8b4e" />

So it's mainly because the u8 version of bilinear rendering works differently, so there's a bigger divergence. I can add a comment for that.

nicoburns commented on sparse_strips/vello_common/src/encode.rs:
```diff
@@ -12,8 +12,10 @@ use crate::paint::{Image, ImageSource, IndexedPaint, Paint, PremulColor};
+#[cfg(not(feature = "multithreading"))]
+use core::cell::OnceCell;
+#[cfg(feature = "multithreading")]
+use once_cell::sync::OnceCell;
```
I think using std::OnceLock is better. It's the same thing as the sync OnceCell from the once_cell crate but avoids bringing in a dependency.

You could import it with an aliases name to avoid having to conditionally compile other code.
